### PR TITLE
Create a Chocolatey Package

### LINKED ISSUE #235: Create Choco Package
Can we have a [chocolatey package](https://chocolatey.org/docs/create-packages) for installing tinytex in windows? I'm not a R user and I'm just asking this for installing latex with ease for other purposes.

Currently the install script in tools directory works good, but yeah it would good for having a choco package by which users install using
```powershell
choco install tinytex
```
Cheers.
---

By filing an issue to this repo, I promise that

- [x] I have fully read the issue guide at https://yihui.org/issue/.
- [x] I have provided the necessary information about my issue.
    - If I'm asking a question, I have already asked it on Stack Overflow or RStudio Community, waited for at least 24 hours, and included a link to my question there.
    - If I'm filing a bug report, I have included a minimal, self-contained, and reproducible example, and have also included `xfun::session_info('tinytex')`. I have upgraded all my packages to their latest versions (e.g., R, RStudio, and R packages), and also tried the development version: `remotes::install_github('yihui/tinytex')`.
    - If I have posted the same issue elsewhere, I have also mentioned it in this issue.
- [x] I have learned the Github Markdown syntax, and formatted my issue correctly.

I understand that my issue may be closed if I don't fulfill my promises.

### PR DESCRIPTION
**Pull request by naveen521kk**
So, here I have an initial choco setup. Need to work on it.
Fixes #235

### CODE REVIEW

**Comment by naveen521kk** [comment]
It's worse. the zip package checksum is changing weirdly and chocolatey doesn't allow without checksum. See the logs
- https://github.com/naveen521kk/tinytex/runs/992776879?check_suite_focus=true#step:3:171
- https://github.com/naveen521kk/tinytex/runs/992762371?check_suite_focus=true#step:3:169

**Comment by naveen521kk** [comment]
@yihui Can you help me what happened [here](https://github.com/naveen521kk/tinytex/runs/1027085152?check_suite_focus=true#step:3:462) ?

**Comment by yihui** [comment]
I don't know why, but the error log indicates that you need to install the package **dehyph-exptl**:

```r
> tinytex::parse_packages(text = "! I can't find file `dehypht-x-2019-04-04.tex'.")
tlmgr search --file --global '/dehypht-x-2019-04-04.tex'
[1] "dehyph-exptl"
```

**Comment by naveen521kk** [comment]
> you need to install the package dehyph-exptl:

Why does this doesn't happen in executing that batch script? Weird. Maybe, I missed something, @yihui , if you are free, can you review whether the `chocolateyinstall.ps1` is nearly same as `install-windows.bat`(the one in tools directory).

**Comment by yihui** [comment]
It is nearly the same.

As I said, I don't know the reason why it failed. By comparing the log on AppVeyor at https://ci.appveyor.com/project/yihui/tinytex/build/job/94ptgpvmyhff0l6s, the only possibility that I could think of is that the two installations used different CTAN mirrors. Sometimes this makes a difference, as one mirror might not be recent enough, or one mirror is too recent to contain bugs.

**Comment by naveen521kk** [comment]
Maybe let me edit the appveyor script also so that it tests this choco package and then test it.

**Comment by naveen521kk** [comment]
Just checking that `tinytex.profile` it generates is correct? @yihui

**Comment by naveen521kk** [comment]
@yihui Now I have enabled running both the script in gh-actions. See logs [here](https://github.com/naveen521kk/tinytex/runs/1030389093?check_suite_focus=true).

**Comment by naveen521kk** [comment]
Yes, I fixed those errors in the latest commits. @yihui Can you review those scripts? 
After that it can be published to chocolatey. Maybe, create a chocolatey account.

File: tools/choco/tinytex/tinytex.nuspec [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+    <projectUrl>https://yihui.org/tinytex/</projectUrl>
+    <iconUrl>https://yihui.org/images/logo-tinytex.png</iconUrl>
+    <copyright>2017-2020 Yihui Xie and RStudio, PBC</copyright>
+    <licenseUrl>https://github.com/yihui/tinytex/blob/master/LICENSE</licenseUrl>
+    <requireLicenseAcceptance>true</requireLicenseAcceptance>
+    <projectSourceUrl>https://github.com/yihui/tinytex/</projectSourceUrl>
+    <docsUrl>https://yihui.org/tinytex/</docsUrl>
+    <bugTrackerUrl>https://github.com/yihui/tinytex/issues</bugTrackerUrl>
+    <tags>tinytex tex tiny</tags>
+    <summary>Tiny Install of Latex</summary>
```
Comment (yihui) [inline tools/choco/tinytex/tinytex.nuspec:21]:
```suggestion
    <summary>Tiny Install of TeX Live</summary>
```

File: tools/choco/tinytex/tinytex.nuspec [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+    <id>tinytex</id>
+    <version>0.25</version>
+    <packageSourceUrl>https://github.com/yihui/tinytex/tree/master/tools/choco/tinytex</packageSourceUrl>
+    <owners>__REPLACE_YOUR_NAME__</owners>
+    <!-- Owner is one who publishes the package to chocolatey. -->
+    <title>Tinytex (Install)</title>
+    <authors>Yihui Xie</authors>
+    <projectUrl>https://yihui.org/tinytex/</projectUrl>
+    <iconUrl>https://yihui.org/images/logo-tinytex.png</iconUrl>
+    <copyright>2017-2020 Yihui Xie and RStudio, PBC</copyright>
```
[the remark this answers is not in the document]
Comment (naveen521kk) [reply tools/choco/tinytex/tinytex.nuspec:14]: I just copied the one in the LICENSE file. Maybe can we create a new LICENSE file inside the `choco` directory? And I will change this to my name.

**Comment by naveen521kk** [comment]
@yihui Where can I document about this?

File: tools/choco/tinytex/tinytex.nuspec (hunk 2 above)
[the remark this answers is not in the document]
Comment (yihui) [reply tools/choco/tinytex/tinytex.nuspec:13]: Sure. It's up to you. As I said, it's mostly your code, so you have the full freedom to choose the license.

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -0,0 +1,34 @@
+name: Get Latest Version And Publish
```
Comment (naveen521kk) [inline .github/workflows/ci.yml:1]: @yihui Do you need this workflow in the repo. I don't think it's necessary. What do you think?

Comment (naveen521kk) [reply .github/workflows/ci.yml]: I removed it. :)

**Comment by naveen521kk** [comment]
@yihui I have pushed the package to chocolatey https://chocolatey.org/packages/tinytex also, I wish to add you also as a maintainer. SO create a [chocolatey account](https://chocolatey.org/account/Register).
Also, merge this after reviewing ASAP, so that they could verify.

**Comment by naveen521kk** [comment]
@yihui Anything to be done before merging this? Did I miss anything?

**Comment by naveen521kk** [comment]
@yihui Could you merge this. Tests in chocolatey package fail there because it couldn't find the url mentioned which would only be done if this merged:(

**Review by yihui — APPROVED** [review]
Sorry I was on vacation. I'll merge the PR in a second. Thanks!

BTW, my account name is `yihui`.

**Comment by yihui** [comment]
BTW, could you correct `TinyTex` to `TinyTeX` on https://chocolatey.org/packages/tinytex? :)

**Comment by naveen521kk** [comment]
I will do tomorrow as soon as I wake up. :)

File: tools/choco/tinytex/tinytex.nuspec [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+    <id>tinytex</id>
+    <version>0.25</version>
+    <packageSourceUrl>https://github.com/yihui/tinytex/tree/master/tools/choco/tinytex</packageSourceUrl>
+    <owners>naveen521kk</owners>
+    <title>TinyTex (Install)</title>
+    <authors>Naveen M K,Yihui Xie</authors>
+    <projectUrl>https://yihui.org/tinytex/</projectUrl>
+    <iconUrl>https://yihui.org/images/logo-tinytex.png</iconUrl>
+    <copyright>Copyright 2020 Naveen M K,Yihui Xie</copyright>
+    <licenseUrl>https://github.com/yihui/tinytex/blob/master/tools/choco/tinytex/LICENSE</licenseUrl>
```
Comment (naveen521kk) [inline tools/choco/tinytex/tinytex.nuspec:14]: I think this should license should be pointed to Texlive license what to do you say @yihui ?

Comment (yihui) [reply tools/choco/tinytex/tinytex.nuspec:14]: Good question. I'm not sure...

Comment (naveen521kk) [reply tools/choco/tinytex/tinytex.nuspec:14]: Ok then let it be.

**Comment by naveen521kk** [comment]
@yihui Do you know whether the checksum of http://mirrors.ctan.org/systems/texlive/tlnet/install-tl.zip can be matched by any chance? They are requesting that. Maybe are we allowed(by TexLive license) to include that zip file inside the package?

**Comment by yihui** [comment]
I don't think it can. It's updated from time to time. I prefer not including install-tl.zip. License is one thing (probably okay). The other thing is that I'm not familiar with the development model/process of TeX Live, so I don't know the possible consequences if we freeze a version of install-tl.zip.

**Comment by naveen521kk** [comment]
@yihui DO you know any place where they have archives of version? Also, what can be done for this? See the review comment and is worse. https://chocolatey.org/packages/tinytex/0.25 and I feel like it looks correct. I had some tries with my PC and the one of my VPS in another place. I get two different checksums in that
`6cac9cbf2c6547db82cb7647e34586c2b2f34431b44a12f081b7430bf8b6237f036b3e01d07f13d936660e4ff0147cea147e2d9a4c4345485d2cf347845f4e98`
and `47c3406fecf8a05b3d7a1ecef52c13d6378b6f7e845df24264837a6ed6f69c7e714e8926ab08f9cc3c59f97038af255eb2e0c0628c87220d33387ef14a35367e` 
both sha512. I really don't know what to do.

**Comment by naveen521kk** [comment]
Or another thing that can be done is include the zip and update the chocolatey package whenever they change it. It can be done automatically, like Github Actions, if I know how can I get whether it changed or not. That's not an efficient idea but is ok. It would be good if we could make up with anybody who knows about it.

**Comment by yihui** [comment]
If we could automate this job of updating the chocolatey package with the latest install-tl.zip, that will be great. I don't know why the checksums are not stable.

There are archives provided by Nobert: https://texlive.info but I'm not sure how "official" this service is.

**Comment by naveen521kk** [comment]
Linking https://tug.org/pipermail/tex-live/2020-September/046053.html for reference. I really don't know why it made the mail as an attachment (lol).
Cc
@yihui

**Comment by yihui** [comment]
Continuing from https://github.com/yihui/tinytex/issues/235#issuecomment-690147866:

So it seems that we have two options:

1. Use the installer from texlive.info, but I don't know how to update it (on a daily basis).

2. Or we pull the zip installer and store it somewhere. I can do that on AppVeyor and give you a link to it. But I don't know where to write the checksum.

To me, 2 seems to be easier, but I'll let you decide.

**Comment by cderv** [comment]
> Or we pull the zip installer and store it somewhere. I can do that on AppVeyor and give you a link to it. But I don't know where to write the checksum.

Using github action , you may be able to write the checksum in a file right in this repository. Just an idea.

**Comment by naveen521kk** [comment]
I have no particular option on that. But I consider the second but with a slight condition. We should be packing the zip file(or maybe an exe self extracting archive) with half build installer(By that I mean we have to just do the `tlmgr` to install packages.). That would save some amount of time on user's PC and could make the install faster.
Alternatively if we don't like to work much we can change the links from CTAN mirrors to just `texlive.info`.(I will work no problem on that ;-)).
> Use the installer from texlive.info, but I don't know how to update it (on a daily basis).

This can be done very easily, see https://github.com/naveen521kk/au-packages, where I maintain a particular set of chocolatey packages which are uploaded once a new version is released. Appveyor, runs daily checking for new version. People at chocolatey made it very easy to set it up and I can do if needed. But I am more towards executables. ;-).

>Or we pull the zip installer and store it somewhere. I can do that on AppVeyor and give you a link to it. But I don't know where to write the checksum.

We can do this and upload to a Github Release along with a file with checksum like for example, a `<filename>.sha256` file with the checksums. Maybe if all the packages in `pkg-custom.txt` are from Texlive then we can pack it fully as how, they do for Debian(I asked Norbert about that and he said so in mail). If we could pack it fully that is best way.

**Comment by naveen521kk** [comment]
@yihui Tell me your choice on that and we can either make an new issue to start working on that.

**Comment by yihui** [comment]
> Using github action , you may be able to write the checksum in a file right in this repository. Just an idea.

@cderv Yes, I can do that, but I was actually asking if the checksum should be written to the spec file of the chocolatey package (`tools/choco/tinytex/tinytex.nuspec`) or somewhere. I don't know how chocolatey verifies the checksum. @naveen521kk certainly knows.

> We should be packing the zip file(or maybe an exe self extracting archive) with half build installer(By that I mean we have to just do the `tlmgr` to install packages.). That would save some amount of time on user's PC and could make the install faster.

@naveen521kk Given Nobert's reply in the mailing list, which implies it is okay to redistribute the TeX Live stuff (installer and packages that belong to TeX Live), I think your idea is great!

> Maybe if all the packages in `pkg-custom.txt` are from Texlive then we can pack it fully as how, they do for Debian(I asked Norbert about that and he said so in mail). If we could pack it fully that is best way.

Now this is exactly my question. Are all packages in `pkg-custom.txt` from TeX Live? I don't know how to verify it. If you do, I'll truly appreciate it if you could do it. If we could just offer a full package, it will save us a huge amount of work in both this chocolatey package and `tinytex::install_tinytex()` (the latter could also just download the prebuilt package, and the installation will be a lot faster). This is the No.1 challenge in TinyTeX that I have been unable to solve, which I said in FAQ 2: https://yihui.org/tinytex/faq/ If we could make sure we won't have licensing problems, I'll be extremely excited to provide prebuilt packages publicly.

**Comment by yihui** [comment]
A little more clarification in case the challenge I mentioned was not clear enough: TinyTeX doesn't include the source files (in order to reduce the package size) https://github.com/yihui/tinytex/blob/2c93cb38cbea6792192d0d40139149d796de8306/tools/tinytex.profile#L9-L10
This was my major concern about redistributing TeX Live and CTAN packages. I don't know much about the TeX Live licence, but GPL is likely to prevent people from distributing GPL-licensed software without providing the source.

One way to remedy this problem is to provide two versions: one with and one without source. Then tell users that if they want the source code, they could download the alternative (bigger) version. I guess this might be fine. I think pretty much everyone uses binary versions of software licensed under GPL, and the source code is certainly not available in the binaries, but users do have access to the source code if they want to. For example, CRAN distributes both source and prebuilt binary versions of R packages.

Anyway, as I said, this problem is too complicated to me. If someone could give me a definitive answer, I'll really appreciate it.

**Comment by naveen521kk** [comment]
> Now this is exactly my question. Are all packages in pkg-custom.txt from TeX Live? I don't know how to verify it. If you do, I'll truly appreciate it if you could do it. 
Ok this is hard to do. I don't know exactly how to do that but I have a rough idea. I think we can get the location of `tar.*` files using `tlmgr`. Look whether it is in `system/texlive` folder(I maybe wrong with name here). If it is then we can include it. If not we should install it later using `tlmgr` as we previous do. @norbusan is it possible for you to help us over here?

> One way to remedy this problem is to provide two versions: one with and one without source. Then tell users that if they want the source code, they could download the alternative (bigger) version.
Its nice one we can do. But isn't it enough to link to them directly as we aren't modifying anything there? Anyways, if we want to make a big one we can do it easily and also we can parameter based installation on chocolatey for that. You would need to handle about the R package.

**Comment by naveen521kk** [comment]
I suggest making a new issue and discussing there would be good.

**Comment by naveen521kk** [comment]
I suggest making a new issue and discussing there would be good.

**Comment by norbusan** [comment]
@naveen521kk What is `pkg-custom.txt` I don't know what I check for?

@yihui concerning source/doc files - in principle it is ok not to distribute the source files and doc files. You can also explain users that by doing `tlmgr installl --reinstall --with-doc --with-src <PKG>` one can get the source and doc files afterwards, too.

**Comment by yihui** [comment]
> `tlmgr installl --reinstall --with-doc --with-src`

@norbusan Oh I didn't know that. That's such an elegant solution! Thank you!

> What is `pkg-custom.txt` I don't know what I check for?

It contains the list of packages to be installed after installing the `infra-only` scheme: https://github.com/yihui/tinytex/blob/master/tools/pkgs-custom.txt

**Comment by norbusan** [comment]
Hi @yihui 
concerning pkg-custom, then I don't understand what the question above:

> Now this is exactly my question. Are all packages in pkg-custom.txt from TeX Live? I don't know how to verify it. If you do, I'll truly appreciate it if you could do it.

should mean?

Of course, everything you install using `tlmgr` is from TeX Live.


THinking about it, I **GUESS** I know the reason for confusion: CTAN serves a *double* purpose, first it contains all the stuff around TeX etc, including non-free etc etc. And, in addition, it serves as distribution media for TeX Live (`/systems/texlive`).

So if `tlmgr` uses the CTAN, then it uses the TeX Live files that are distributed via CTAN.

So to repeat, everything you install via tlmgr from a TeX Live repo, that is including the CTAN mirrors of the TeX LIve repo, is *from* TeX Live and thus redistributable.

But if you go to `CTAN/tex-archive/....` there you might find things that are not redistributable.

Does this clear up the confusion a bit?

**Comment by naveen521kk** [comment]
I think yes @norbusan it answers my question. Now the work is to make a package for TinyTeX to create some executables. Before you told this I tried something worse by checking http://mirror.ctan.org/systems/texlive/tlnet/archive/ for all packages in pkg-customs.txt using a python program. I was *Stupid*. (Everything was there).


And we would be having no problem with the TexLive license. @yihui let's working on setting up Github Actions to run it. I will make a PR soon.  I probably gonna create a "classic" installer using NSIS.

**Comment by yihui** [comment]
@norbusan Thanks! That was extremely helpful!

**Pull request merged**